javaslang/src/main/java/javaslang/collection/BitMappedTrie.java

```diff
@@ -131,11 +134,33 @@ private static int treeSize(int branchCount, int depthShift) {
         } else if (n <= 0) {
             return empty();
         } else {
-            final Object[] root = modifyLeaf(array, depthShift, (offset + n) - 1, Arrays::copyTake, Arrays::copyTake);
-            return new BitMappedTrie<>(root, offset, n, depthShift);
+            final int index = n - 1;
+            final Object[] root = arePointingToSameLeaf(index, length() - 1)
+                                  ? array
+                                  : modifyLeaf(array, depthShift, offset + index, Arrays::copyTake, IDENTITY);
+            return collapsed(root, offset, n, depthShift);
         }
     }
 
+    private boolean arePointingToSameLeaf(int i, int j) {
+        final boolean result = firstDigit(offset + i, BRANCHING_BASE) == firstDigit(offset + j, BRANCHING_BASE);
+        assert result == (getLeaf(i) == getLeaf(j));
+        return result;
+    }
+
+    /* drop root node while it has a single element */
+    private static <T> BitMappedTrie<T> collapsed(Object[] array, int offset, int length, int shift) {
```

Comment:
we could also drop leading `null`s, but that wouldn't result in any speed advantage, just aesthetics...